pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,278 @@
+				topologyv1alpha1.PartitionSetValid,
+				topologyv1alpha1.PartitionSetInvalidSelectorReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				err.Error(),
+			)
+			conditions.MarkFalse(
+				partitionSet,
+				topologyv1alpha1.PartitionsReady,
+				topologyv1alpha1.ErrorGeneratingPartitionsReason,
+				conditionsv1alpha1.ConditionSeverityError,
+				"",
+			)
+			// Deleting existing partitions when the new selector is invalid
+			if err = c.deletePartitions(ctx, partitionSet); err != nil {
+				return err
+			}
+			// No need to requeue if the selector is not valid
+			return nil
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionSetValid)
+
+	shards, err := c.listShards(labelSelector)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error listing Shards",
+		)
+		return err
+	}
+
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(partitionSet)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"error generating PartitionSet key",
+		)
+		return err
+	}
+	oldPartitions, err := c.getPartitionsByPartitionSet(key)
+	if err != nil {
+		conditions.MarkFalse(
+			partitionSet,
+			topologyv1alpha1.PartitionsReady,
+			topologyv1alpha1.ErrorGeneratingPartitionsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"unable to retrieve associated Partitions",
+		)
+		return err
+	}
+
+	var matchLabelsMap map[string]map[string]string
+	if partitionSet.Spec.ShardSelector != nil {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, partitionSet.Spec.ShardSelector.MatchLabels)
+	} else {
+		matchLabelsMap = partition(shards, partitionSet.Spec.Dimensions, nil)
+	}
+	partitionSet.Status.Count = uint16(len(matchLabelsMap))
+	existingMatches := map[string]struct{}{}
+	var placeHolder struct{}
+	newMatchExpressions := []metav1.LabelSelectorRequirement{}
+	if partitionSet.Spec.ShardSelector != nil {
+		newMatchExpressions = partitionSet.Spec.ShardSelector.MatchExpressions
+	}
+	for _, oldPartition := range oldPartitions {
+		// MatchExpressions need to be the same
+		oldMatchExpressions := []metav1.LabelSelectorRequirement{}
+		if oldPartition.Spec.Selector != nil {
+			oldMatchExpressions = oldPartition.Spec.Selector.MatchExpressions
+		}
+		if !equality.Semantic.DeepEqual(oldMatchExpressions, newMatchExpressions) {
+			if err := c.deletePartition(ctx, logicalcluster.From(oldPartition).Path(), oldPartition.Name); err != nil && !apierrors.IsNotFound(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
```

Comment:
again, since this is a network call, we could be more greedy and try all `oldPartitions`, wdyt?

Reply:
I am not sure how you would do that here. There is no single selector for the Partitions to delete.